### DISCUSSION TITLE
Working on keybindings editor

### DISCUSSION THREAD

**Discussion by MrApplejuice** [mbrlabs/Lorien#144] [General]
**_As you (maintainers) asked for, I am opening this discussion here, so we can discus a bit more how this should be implemented._**

First of all: I really really love to have found this application. Infinite canvas on Linux - I was looking for something like the [Concepts app](https://concepts.app/en/) on Linux for ages! A very nice start and a good idea to use a game-engine for this!

Anyway, I found a problem that I started to tackle already mentioned in here: https://github.com/mbrlabs/Lorien/issues/134 . Without keybindings I cannot make use of my 12 custom buttons on my graphics tablet, making zooming and selecting tools unnecessarily cumbersome.

I am fairly familiar with Godot and cobbled something together yesterday already, but came to the conclusion that this will likely grow into a "big feature":

![grafik](https://user-images.githubusercontent.com/542105/173515390-a406fb04-3fff-4417-8381-8388c302d6cf.png)

# Problems

The keybindings editor right now is implemented in a quite simple way:

- Enumerate all actions configured
- Load translations for all action items
- Allow to rebind keyboard shortcuts to each action

However, there are open issues:
- The keybindings need to be preserved - probably as part of the overall settings
- *Big one*: Lorien right now has the keybindings encoded in the translation strings themselves

~~~~~~~~~
TOOLBAR_TOOLTIP_NEW_FILE        Create New File (CTRL+N)
TOOLBAR_TOOLTIP_OPEN_FILE       Open File (CTRL+O)
TOOLBAR_TOOLTIP_SAVE_FILE       Save File (CTRL+S)
TOOLBAR_TOOLTIP_UNDO            Undo Action (CTRL+Z)
TOOLBAR_TOOLTIP_REDO            Redo Action (CTRL+Y)
TOOLBAR_TOOLTIP_BRUSH_TOOL      Brush Tool (B)
~~~~~~~~~

If the keybindings are editable, this does not work anymore. I discuss my proposal for the solution to this below.

# Solving keybinding lists in translation strings

There are multiple steps to this. First of all, editing keybindings should probably not require restarting the application. That would be very annoying. But the strings for tooltips (and menu items?) still should update, I think. I `grep`ped the codebase for use of the TranslationServer and I think the main reason why live-changing of the language does not work are a _few_ instances where "tr()" is used manually, am I right?

If that is the case, I could fix this by emitting a global-event that indicates that the translation table has updated and fix live-language switching that way. 

Fixing live-language switching would then allow me to use the translation-engine itself to do some basic string interpolation and change, for example the toolbar translations to something like:

~~~~~~~~~
TOOLBAR_TOOLTIP_NEW_FILE        Create New File {{keybindings:ACTION_new_file}}
TOOLBAR_TOOLTIP_OPEN_FILE       Open File {{keybindings:ACTION_open_file}}
[...]
~~~~~~~~~

The `{{...}}`-bit would be a basic parser that allows Lorien to do some advanced string manipulations, similar to how [Python's jinja2 library](http://jinja.octoprint.org/) does it (but waaay simpler). I do not want to write a full-Domain Specific Language parser for this, but something simple with regexes. When keybindings update, I can then update the entire translation table and auto-update the keybinding lists and apply the new tooltips etc. that way.

# Plan (order of PRs I propose to get to this feature)

1. I create a patch so that live-updating of the language works
2. I introduce the `Create New File {{keybindings:ACTION_new_file}}` interpolation for the translation files
3. I continue work on the keybindings editor and make use the previous two additions 

Splitting the work packages up like this should prevent a "mega PR" from happening...

Does that sound like a sound plan?

Kind regards,
  Paul K.

--------------------------
Edit: Here my initial work https://github.com/MrApplejuice/Lorien/tree/keybindings

**Comment by MrApplejuice** [comment]
Offtopic but related: I think it might be worth while (in the future) to replace the current text-files that are used for translations. While simple, they are less standard and likely will cause confusion among other contributors in the long run. I would maybe consider moving to [gettext](https://www.gnu.org/software/gettext/manual/html_node/index.html) afterall. It also allows this project to use some other cool external tools for translations like https://docs.weblate.org/en/latest/ or others to make it easier to update translation strings.

Anyway, just my thoughts :-)

**Reply by mbrlabs** [reply]
You're not the first one who suggested Weblate: https://github.com/mbrlabs/Lorien/issues/70

I'm going to look into it eventually ;)

**Comment by mbrlabs** [comment]
Sounds like a great plan! Yes, splitting a big change like this up into muliple PRs makes it much easier for me to review it etc. 👍

> the main reason why live-changing of the language does not work are a few instances where "tr()" is used manually, am I right?

Yes, correct. But you fixed that already anyway :)

**Comment by MrApplejuice** [comment]
@mbrlabs I have a small problem - multiple solutions, but I want to propose something that could be called a "scary change".

# Problem

For the keybindings-dialog I was planning to implement a "press button to set shortcut" feature. The problem with that feature is that I, of course, do not know what keystrokes a user might press. Therefore, I need to suppress the shortcut detection application wide.

My normal go-to solution for this is to filter events in the `Node._input`-tree of the event processing using `ScreneTree.set_input_handled()` to prevent further processing of an input event. However, I discovered that that does not work in Lorien, because Lorien uses a gameloop-style shortcut detection as part of the `_process()`-tree, which does not allow me to suppress the event.

# "Scary" (but cleanest) solution

I propose to switch the InputEvent-handling in Lorien `Main.gd` to use `_unhandled_input(event)` instead of the `_process()`-callback. Here the suggested changes for the lines in Main.gd:

https://github.com/mbrlabs/Lorien/blob/bc83fbe6691c0879335d0f2577443bad076f47f3/lorien/Main.gd#L116-L133

Should become.

~~~~~~~~~~~~.gdscript
# -------------------------------------------------------------------------------------------------
func _process(delta):
	# _handle_input_actions()
	_statusbar.set_stroke_count(_canvas.info.stroke_count)
	_statusbar.set_point_count(_canvas.info.point_count)
	_statusbar.set_pressure(_canvas.info.current_pressure)
	_statusbar.set_camera_position(_canvas.get_camera_offset())
	_statusbar.set_camera_zoom(_canvas.get_camera_zoom())
	_statusbar.set_fps(Engine.get_frames_per_second())
	
	# Update tab title
	var active_project: Project = ProjectManager.get_active_project()
	if active_project != null:
		_menubar.update_tab_title(active_project)

# -------------------------------------------------------------------------------------------------
func _input(event):
	# Allows event filtering using: get_tree().set_input_as_handled()
	pass

# -------------------------------------------------------------------------------------------------
# func _handle_input_actions():
func _unhandled_input(event):
	if !is_dialog_open():
		var copy := Input.is_action_just_pressed("copy_strokes")
~~~~~~~~~~~~

This would switch input processing around and and to event-based input-event processing instead of _process-loop driven input event processing. I think that this is the suggested way to do things anyway according to the docs?

I think this workaround for detecting if a modal UI element is open would also not be necessary anymore because of the[ order that event are processed in the event loop](https://docs.godotengine.org/en/stable/tutorials/inputs/inputevent.html?highlight=handle%20event#how-does-it-work):

~~~~~~~~~~~~~~
func _unhandled_input(event):
	if !is_dialog_open():
~~~~~~~~~~~~~~

I did a quick check and everything seems to still work. However, I do not know if there are any deeper reasons to why input processing is done the way it is? If there is, there are a lot of other ways that I can implement to suppress input event processing for the keybindings editor, but I think the suggested change is the cleanest way forward...

**Reply by mbrlabs** [reply]
I don't think there was any reason in particular. I did a quick test myself and did not notice any issues, so i'd say go ahead.

**Comment by HuntClauss** [comment]
I don't know if this is the case in your solution, but when I was playing with custom shortcuts I occur a problem with overlapping keybinds like Control+Z and Control+Shift+Z

**Reply by MrApplejuice** [reply]
What is the problem exactly? That should in principle work just fine, right? Or do you mean that there is a problem _right now_?

If so, I am expecting you found a event-processing bug that I am currently trying to sort out as well: I found that a lot of key-events are being matched in the main-event in parallel. The could mean that you could trigger Control+Shift+Z and Control+Z at once by pressing "Control+Shift+Z" . I assume that is what is happening? I am facing this problem with the keybinding editor where one should be able to hit _any_ key, but that also triggers some background-keyevents right now which is unintended.

I am currently trying to rewrite the event processing to use event-processing to fix this issue, maybe also the one you are facing? But I have hit a few roadblocks with the used ViewportContainer. It will probably take a little longer until that is fixed, but progress is quite good right now :-)
